### PR TITLE
[WIP] implemented file-download-barrier

### PR DESCRIPTION
**Pull request by georgiy-abbasov**

### CODE REVIEW

File: src/client/driver/driver.js [hunk 1]
```diff
@@ -132,19 +130,25 @@ export default class Driver {
...
-        if (this.beforeUnloadRaised)
-            return hangingPromise;
 
-        return transport
-            .queuedAsyncServiceMsg({ cmd: TEST_RUN_MESSAGES.ready, status })
-            .then(res => {
-                //NOTE: do not execute the next command if the page is unloading
-                if (this.beforeUnloadRaised)
-                    return hangingPromise;
+        var checkForUnloading = () => {
```
Comment (AlexanderMoskovkin) [inline src/client/driver/driver.js]: pageUnloadBarrier.wait(0)

File: src/client/runner/step-iterator.js [hunk 2]
```diff
@@ -43,7 +43,8 @@ var StepIterator = function (pingIframe) {
     this.globalWaitForEvent   = null;
     this.globalWaitForTimeout = null;
 
-    this.eventEmitter = new serviceUtils.EventEmitter();
+    this.eventEmitter      = new serviceUtils.EventEmitter();
+    this.pageUnloadBarrier = pageUnloadBarrier;
```
Comment (AlexanderMoskovkin) [inline src/client/runner/step-iterator.js]: delete this line

File: src/client/runner/step-iterator.js [hunk 3]
```diff
@@ -233,8 +234,8 @@ StepIterator.prototype._completeAsyncAction = function () {
     if (iterator.state.stopped)
         return;
 
-    pageUnloadBarrier
-        .wait()
+    this.pageUnloadBarrier
```
Comment (AlexanderMoskovkin) [inline src/client/runner/step-iterator.js]: revert this change

File: src/test-run/index.js [hunk 4]
```diff
@@ -90,7 +91,7 @@ export default class TestRun extends Session {
     }
 
     handleFileDownload () {
-        // TODO
+        this.isFileDownloading = true;
```
Comment (AlexanderMoskovkin) [inline src/test-run/index.js]:
```js
if(this.resolveWaitForFileDownloadingPromise)
    this.resolveWaitForFileDownloadingPromise
``

File: src/test-run/index.js [hunk 5]
```diff
@@ -343,3 +344,15 @@ ServiceMessages[CLIENT_MESSAGES.readyForBrowserManipulation] = async function (m
...
+    var isFileDownloading = this.isFileDownloading;
+
+    this.isFileDownloading = false;
+
+    return isFileDownloading;
+};
+
+ServiceMessages[CLIENT_MESSAGES.uncheckFileDownloadingFlag] = function () {
+    this.isFileDownloading = false;
+};
```
Comment (AlexanderMoskovkin) [inline src/test-run/index.js]:
```js
ServiceMessgae[CLIENT_MESSAGES.waitForFileDownloading] = function () {
    return Promise(resolve => this.resolveWaitForFileDownloadingPromise = resolve);
};
```

File: src/client/core/page-unload-barrier.js [hunk 6]
```diff
@@ -84,11 +84,11 @@ export function init () {
     handleBeforeUnload();
 }
 
-export function wait () {
+export function wait (options) {
     return new Promise(resolve => {
         delay(DEFAULT_BARRIER_TIMEOUT)
             .then(() => {
-                if (unloading)
+                if (options && options.isLegacy && unloading)
```
Comment (AlexanderMoskovkin) [inline src/client/core/page-unload-barrier.js]:
```js
if(unloading) {
waitForFileDownloading.then(resolve);
return;
}
```

File: src/client/core/page-unload-barrier.js [hunk 7]
```diff
@@ -84,11 +84,11 @@ export function init () {
...
                     return;
 
                 if (!waitingForUnload)
```
Comment (AlexanderMoskovkin) [inline src/client/core/page-unload-barrier.js]: check why waitingForUnload not set, write the test

**Comment by AlexanderMoskovkin** [comment]
waiting for https://github.com/DevExpress/testcafe-hammerhead/issues/667

**Comment by georgiy-abbasov** [comment]
@AlexanderMoskovkin, i think, there is no need for this approach anymore (cause we merged https://github.com/DevExpress/testcafe/pull/759, separeted legacy and next-api, ). I think we should open separate PR when new `beforeUnload` will be ready (we can take tests for it from this PR).

**Comment by AlexanderMoskovkin** [comment]
yep
